### PR TITLE
Add code to find branchings and arborescences.

### PR DESCRIPTION
**Pull request by chebee7i**
This is *old* code, but it is rotting on my hard drive. So I figured it would be better to try and merge it. I don't doubt that this could could eventually be cleaned up and optimized, but it works and has unit tests right now. 

Possible refactorings are that it could make use of the functionality provided by https://github.com/networkx/networkx/pull/1314.  Unique to this algorithm is that it requires keys for each edge and as part of the algorithm, edges are modified with new heads and tails, but the identity of the edge is unchanged. It would benefit greatly from a subclass of MultiDiGraph that tracked the edges by a graph-wide edge key.

Anyway, it provides code for finding optimium (min or max) branchings and arborescences. These are the directed analogs of min/max spanning trees on undirected graphs.

### CODE REVIEW

**Comment by hagberg** [comment]
Thanks for uploading this.  I see what you mean about the global edge key.
I suggest we merge.

**Comment by chebee7i** [comment]
So strange. I'm not getting any Python 3 errors now, but the one on TravisCI doesn't seem so unsurprising either.

File: networkx/algorithms/tree/branchings.py [hunk 1]
```diff
@@ -0,0 +1,659 @@
...
+                        minweight = np.inf
+                        minedge = None
+                        Q_incoming_weight = {}
+                        for edge_key in Q_edges:
+                            u, v, data = B.edge_index[edge_key]
+                            w = data[attr]
+                            Q_incoming_weight[v] = w
+                            if  w < minweight:
+                                minweight = w
+                                minedge = edge_key
```
Comment (hagberg) [inline networkx/algorithms/tree/branchings.py]: The print here will break python3 tests.

Comment (chebee7i) [inline networkx/algorithms/tree/branchings.py]: Thanks. Also had to adjust the UnionFind class to handle mixed types.

**Comment by hagberg** [comment]
It looks like this is just failing on the numpy import with pypy.  But I only see numpy used for numpy.inf.
Could it be replaced with float('inf') and drop the numpy dependency?

**Comment by chebee7i** [comment]
Someday, I'll get to adding much more extensive documentation and examples (but not in this PR). I've got one other thing I need to investigate before this can be "done".

**Comment by chebee7i** [comment]
Ok, ready for more review/merging.

**Pull request merged**